File: browser/src/matte.ts
// 去背 — 对应 Python core/matte.py。统一用 floodBg(四角种子连通域), 不是简单色键。
// 全在 RGBA buffer 上操作(原地改 alpha, RGB 保留), 与 Canvas 无关 → 可合成测。
import type { RGBA } from './model';

/** 四角种子 + 区域生长连通域去背。邻居与【当前像素】色差≤tol 才并入 → 顺背景渐变一路走,
 *  撞角色硬边(大跳变)即停, 角色孤岛不连角落即安全。对应 matte.flood_bg。 */
export function floodBg(img: RGBA, tol = 45): RGBA {
  const { data, width: w, height: h } = img;
  if (w < 2 || h < 2) return img;
  const t2 = tol * tol;
  const seen = new Uint8Array(w * h);
  const qx: number[] = [], qy: number[] = [];
  for (const [sx, sy] of [[0, 0], [w - 1, 0], [0, h - 1], [w - 1, h - 1]] as const) {
    const i = sy * w + sx;
    if (!seen[i] && data[i * 4 + 3] !== 0) { seen[i] = 1; qx.push(sx); qy.push(sy); }
  }
  let head = 0;
  while (head < qx.length) {
    const x = qx[head]!, y = qy[head]!; head++;
    const ci = (y * w + x) * 4;
    const cr = data[ci]!, cg = data[ci + 1]!, cb = data[ci + 2]!; // 当前像素色(读后再清 alpha)
    data[ci + 3] = 0;
    for (const [nx, ny] of [[x + 1, y], [x - 1, y], [x, y + 1], [x, y - 1]] as const) {
      if (nx >= 0 && nx < w && ny >= 0 && ny < h) {
        const ni = ny * w + nx;
        if (!seen[ni]) {
          const pi = ni * 4;
          const dr = data[pi]! - cr, dg = data[pi + 1]! - cg, db = data[pi + 2]! - cb;
          if (data[pi + 3]! && dr * dr + dg * dg + db * db <= t2) {
            seen[ni] = 1; qx.push(nx); qy.push(ny);
          }
        }
      }
    }
  }
  return img;
}

/** 色键去背(辅助): 与 key 色差≤tol 的不透明像素转透明。抠洋红标记线 / 手动选色补刀。
 *  对应 matte.key_out。 */
export function keyOut(img: RGBA, key: readonly [number, number, number], tol: number): RGBA {
  const { data } = img;
  const [kr, kg, kb] = key;
  for (let i = 0; i < data.length; i += 4) {
    if (data[i + 3]! &&
        Math.abs(data[i]! - kr) <= tol &&
        Math.abs(data[i + 1]! - kg) <= tol &&
        Math.abs(data[i + 2]! - kb) <= tol) {
      data[i + 3] = 0;
    }
  }
  return img;
}

/** 去绿幕溢出: 不透明像素若 green 明显高于 red/blue, 把 green 压到 max(r,b) → 中和边缘残绿。
 *  对应 matte.despill_green (Python int() 截断 → Math.trunc)。 */
export function despillGreen(img: RGBA, amount = 1.0): RGBA {
  const { data } = img;
  for (let i = 0; i < data.length; i += 4) {
    const r = data[i]!, g = data[i + 1]!, b = data[i + 2]!;
    if (data[i + 3]! && g > r && g > b) {
      const cap = Math.max(r, b);
      data[i + 1] = Math.trunc(g - (g - cap) * amount);
    }
  }
  return img;
}

function median(xs: number[]): number {
  const s = [...xs].sort((a, b) => a - b);
  const m = s.length >> 1;
  return s.length % 2 ? s[m]! : (s[m - 1]! + s[m]!) / 2;
}

/** 四角中位背景色 + 全图色键: 自适应任何纯色幕, 还吃掉被角色包住、连不到四角的背景口袋
 *  (手指缝/手臂弯)。对应 matte.flood_key。 */
export function floodKey(img: RGBA, tol = 80): RGBA {
  const { data, width: w, height: h } = img;
  if (w < 2 || h < 2) return img;
  const cs: number[] = [], gs: number[] = [], bs: number[] = [];
  for (const [cx, cy] of [[0, 0], [w - 1, 0], [0, h - 1], [w - 1, h - 1]] as const) {
    const i = (cy * w + cx) * 4;
    if (data[i + 3] !== 0) { cs.push(data[i]!); gs.push(data[i + 1]!); bs.push(data[i + 2]!); }
  }
  if (!cs.length) return img;
  const cr = median(cs), cg = median(gs), cb = median(bs);
  const t2 = tol * tol;
  for (let i = 0; i < data.length; i += 4) {
    if (data[i + 3]!) {
      const dr = data[i]! - cr, dg = data[i + 1]! - cg, db = data[i + 2]! - cb;
      if (dr * dr + dg * dg + db * db <= t2) data[i + 3] = 0;
    }
  }
  return img;
}

File: core/matte.py
"""core/matte.py — 去背 (单一权威实现, ARCHITECTURE.md §1 / 关键不回归点 2)

去背统一用 flood_bg (四角种子 + 区域生长连通域), 不是简单色键。
key_out (色键) 仅作辅助: 抠导出标记的纯洋红线 + 用户手动选色补刀。
"""
from __future__ import annotations
from collections import deque
from typing import Any, Sequence
from PIL import Image


def flood_bg(img: Image.Image, tol: int = 45) -> Image.Image:
    """连通域去背 (FrameRonin 思路 + 区域生长强化, 四角种子):
    从四角漫水, 邻居与【当前像素】色差≤tol 才并入 → 顺着背景渐变一路走(渐变绿/蓝深浅全吃),
    撞到角色硬边(大跳变 >tol)即停, 不误伤角色; 角色是孤岛不连角落, 身上有同色也安全。
    比"邻居比固定种子"更扛渐变(种子法漫到半路色差超阈就断, 留中间带)。按格调用穿过洋红格线。
    """
    img = img.convert('RGBA'); w, h = img.size
    if w < 2 or h < 2:
        return img
    px: Any = img.load()  # RGBA 已转, 像素访问类型按模式动态 → Any (PIL stub 把 load() 标 Optional+联合, 此处恒非 None)
    t2 = tol * tol
    seen = bytearray(w * h); dq: deque[tuple[int, int]] = deque()
    for sx, sy in ((0, 0), (w-1, 0), (0, h-1), (w-1, h-1)):
        i = sy * w + sx
        if not seen[i] and px[sx, sy][3] != 0:
            seen[i] = 1; dq.append((sx, sy))
    while dq:
        x, y = dq.popleft()
        cr, cg, cb, _ = px[x, y]       # 当前像素色(读后再清, RGB 保留)
        px[x, y] = (cr, cg, cb, 0)
        for nx, ny in ((x+1, y), (x-1, y), (x, y+1), (x, y-1)):
            if 0 <= nx < w and 0 <= ny < h:
                ni = ny * w + nx
                if not seen[ni]:
                    pr, pg, pb, pa = px[nx, ny]
                    if pa and (pr-cr)**2 + (pg-cg)**2 + (pb-cb)**2 <= t2:
                        seen[ni] = 1; dq.append((nx, ny))
    return img


def key_out(img: Image.Image, key: Sequence[int], tol: int) -> Image.Image:
    """色键去背 (辅助): 把与 key 色差 ≤tol 的不透明像素转透明。
    用途: 抠导出横图的纯洋红标记线 (MARK), 或用户手动选色补刀。RGB 保留只清 alpha。
    """
    img = img.convert('RGBA')
    px: Any = img.load(); w, h = img.size  # 像素访问按模式动态 → Any
    kr, kg, kb = key
    for y in range(h):
        for x in range(w):
            r, g, b, a = px[x, y]
            if a and abs(r-kr) <= tol and abs(g-kg) <= tol and abs(b-kb) <= tol:
                px[x, y] = (r, g, b, 0)
    return img


def despill_green(img: Image.Image, amount: float = 1.0) -> Image.Image:
    """去绿幕溢出 (绿底抠像收尾): 不透明像素若 green 明显高于 red/blue, 把 green 压到
    max(r,b) → 中和角色边缘那圈"角色色和绿混"的残留绿 (flood_bg 抠不掉的半透明边缘),
    既不留绿边、也不像白描边那样给角色加一圈白。角色身上避开纯绿时几乎不动本色。
    """
    img = img.convert('RGBA')
    px: Any = img.load(); w, h = img.size  # 像素访问按模式动态 → Any
    for y in range(h):
        for x in range(w):
            r, g, b, a = px[x, y]
            if a and g > r and g > b:
                cap = max(r, b)
                px[x, y] = (r, int(g - (g - cap) * amount), b, a)
    return img


def flood_key(img: Image.Image, tol: int = 80) -> Image.Image:
    """纯色幕去背 (四角采样背景色 + 全图色键): 取**四角中位背景色**(抗噪) → 自适应任何纯色幕
    (绿/蓝/洋红/灰…, 换模型换底色都不用改); 然后**全图扫一遍, 把所有"接近背景色"的像素抠掉**
    ——不只大面积背景, 还包括**被角色包住、连不到四角的背景口袋**(手指缝/手臂弯/裤裆等,
    连通域漫水进不去会漏绿)。纯色幕 + 角色身上避开幕色(提示词已要求)时安全。
    tol 是固定背景容差(任何纯色幕通用, 不用按图调); 边缘抗锯齿过渡由 despill_green 收尾中和。
    """
    import statistics
    img = img.convert('RGBA'); w, h = img.size
    if w < 2 or h < 2:
        return img
    px: Any = img.load()  # 像素访问按模式动态 → Any
    corners = [p for p in (px[0, 0], px[w-1, 0], px[0, h-1], px[w-1, h-1]) if p[3] != 0]
    if not corners:
        return img
    cr = statistics.median(c[0] for c in corners)
    cg = statistics.median(c[1] for c in corners)
    cb = statistics.median(c[2] for c in corners)
    t2 = tol * tol
    for y in range(h):
        for x in range(w):
            r, g, b, a = px[x, y]
            if a and (r-cr)**2 + (g-cg)**2 + (b-cb)**2 <= t2:
                px[x, y] = (r, g, b, 0)
    return img
